### PR TITLE
WordPress Workbench

### PR DESCRIPTION
**Pull request by iliajie**
G'day y'all! ☺️ 

I'm excited to present a new **WordPress Workbench** for discussion and review! Inspired by Joe’s brilliant idea, this tool is an attempt to tap into WordPress’s massive user base, making it a powerful way to attract more users to Virtualmin Pro.

I've implemented most of the features, but I still have a few more ideas in mind that I’ll be adding as more commits to this PR.

I'm starting now, as I’d love to get things moving!

Jamie, below you will find some self-explanatory screenshots showcasing the current functionality. One of the key features I’m still working on is caching. The "Clone" tab also needs a bit more work. Everything else has been tested multiple times and is working smoothly.

Before the final release, I’ll _probably_ add an extra "Dashboard" tab to give an overview of the essentials of the given WordPress instance. However, this is a moot point, as it may not be very practical.

This update also introduces an API that can be used by other web app installers to manage other applications in similar way.

In parallel, I'm submitting the second part of the PR in the Virtualmin Pro repo, which handles the actual submissions.

Anyone else reading this—the work is about 90% done, but I’m still open to suggestions and feedback, so feel free to share any ideas!

All features have a tooltip that explains in detail what they are and what they do.​

Anyone willing to give it a try will need a nightly version of Webmin to get all the latest changes on the Authentic theme side.​

#### Screenshots

Screenshot 2024-09-05 at 19 59 55
Screenshot 2024-09-05 at 19 59 37
Screenshot 2024-09-05 at 20 00 32
image

Screenshot 2024-09-05 at 20 00 53
Screenshot 2024-09-05 at 20 01 15
Screenshot 2024-09-05 at 20 01 47
image

Screenshot 2024-09-05 at 20 02 23
Screenshot 2024-09-05 at 19 58 52
Screenshot 2024-09-05 at 20 02 45
Screenshot 2024-09-05 at 20 02 55
image

### CODE REVIEW

**Comment by pixel-paul** [comment]
Looks great @iliajie

Is there a better place to discuss features, or is the PR the only place?

**Comment by iliajie** [comment]
> Is there a better place to discuss features, or is the PR the only place?

We can discuss it here.

**Comment by pixel-paul** [comment]
1. Have you considered support for mu-plugins? Being able to add a plugin at install that cannot be deactivated would be very useful.

2. Do the backups run like a virtualmin backup? So can be scheduled etc?

3. I'm intrigued about the caching feature!

**Comment by jcameron** [comment]
@iliajie is there a reason why this is entirely implemented in the Virtualmin core? I thought in past discussions we talked about it being a plugin..

**Comment by iliajie** [comment]
> 1. Have you considered support for mu-plugins?

No, not really. There are many downsides to mu-plugins. I don't think we should support it.

>2. Do the backups run like a virtualmin backup? So can be scheduled etc?

Not currently, backups yet cannot be scheduled.

**Comment by iliajie** [comment]
> is there a reason why this is entirely implemented in the Virtualmin core

I'm sorry, what was "entirely implemented" in the Virtualmin core?

If you're referring to it being part of the **Manage Web App** page, then yes, that's expected, as any web app can have this kind of functionality.

> I thought in past discussions we talked about it being a plugin..

It seems there may have been a misunderstanding, because this functionality is strictly specific to the web apps, not Virtualmin or its plugins.

**Comment by jcameron** [comment]
Oh I thought initially it was going to be a plugin ... but it's fine either way.

So is this code done yet? I still am seeing some updates, so I didn't want to do the full code review yet..

**Comment by iliajie** [comment]
> So is this code done yet? I still am seeing some updates, so I didn't want to do the full code review yet..

Not 100%. I was planning to finish today, but we all caught a virus, so I’m feeling under the weather right now. I’ll ping you when it’s ready, hopefully later today.

**Comment by jcameron** [comment]
Ok let me know when it'sr eady for review!

**Comment by iliajie** [comment]
> Ok let me know when it'sr eady for review!

The initially announced features are now implemented, including backups, cloning, Fail2Ban protection and all other that can be found on the UI.

I haven't added caching yet; while caching could help when collecting admin users on slower systems, I’ve tested it with databases of over 100k users, and the performance difference was insignificant, around 250ms without any noticeable lag. Though, I may consider adding it in the future, along with a new "Dashboard" tab.

At this point, I’d like to move on with the review. I’ve poured a lot into getting these core features up and running and, after being sick this past week, I’m not ready to dive back into development just yet. I’ll get there eventually, likely before the final release, but for now, let's focus on the review.

File: virtual-server-lib-funcs.pl [hunk 1]
```diff
@@ -12945,17 +12945,17 @@ sub compression_format
 return $rv;
 }
 
-# extract_compressed_file(file, [destdir])
+# extract_compressed_file(file, [destdir], [&dom])
```
Comment (jcameron) [inline virtual-server-lib-funcs.pl:12948]: This could also be a separate PR

File: workbench.cgi [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#!/usr/local/bin/perl
+# Script workbench
+# workbench.cgi
+
+use File::Basename;
+require './virtual-server-lib.pl';
+&ReadParse();
+# Checks
+&error_setup($text{'scripts_ekit'});
+my ($d, $p) = (&get_domain($in{'dom'}), 'pro');
```
Comment (jcameron) [inline workbench.cgi:10]: Why does $p need to be a variable if it's always 'pro' ?

File: workbench.cgi [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+# Get script
+my ($sinfo) = grep { $_->{'id'} eq $in{'sid'} } &list_domain_scripts($d);
+my $script = &get_script($sinfo->{'name'});
+$script || &error($text{'scripts_emissing'});
+# Error message
+&error("@{[&text('scripts_gpl_pro_tip_workbench_pro_only', $script->{'desc'})]}
+        @{[&text('scripts_gpl_pro_tip_enroll_single', $virtualmin_shop_link)]}")
+                if (defined($in{$p}) && $in{$p} ne $virtualmin_pro);
+# Run
+my $lib = "./$p/scripts/@{[ basename($0) =~ /^(.*)\.cgi$/ ]}-lib.pl";
```
Comment (jcameron) [inline workbench.cgi:22]: Doesn't this `@{[ basename($0) =~ /^(.*)\.cgi$/ ]}` always just evaluate to `workbench` ?

File: edit_phpmode.cgi [hunk 4]
```diff
@@ -176,15 +168,15 @@ if ($canv && !$d->{'alias'} && $mode ne "mod_php") {
 		# System has only one version
 		$fullver = $avail[0]->[1] ? &get_php_version($avail[0]->[1], $d)
 					  : $avail[0]->[0];
-		print &ui_table_row($text{'phpmode_version'}, $fullver . &$get_php_info_bubble('label'))
+		print &ui_table_row($text{'phpmode_version'}, $fullver . &get_php_info_link($d->{'id'}, 'label'))
```
[the remark this answers is not in the document]
Comment (iliajie) [reply edit_phpmode.cgi]: Yes, you're right! This change is unrelated! Fixed [here](https://github.com/virtualmin/virtualmin-gpl/pull/896/commits/7e56ee2c210f240d772ad47df574b3ceab8e2408) and created a [separate PR](https://github.com/virtualmin/virtualmin-gpl/pull/903) for this.

File: virtual-server-lib-funcs.pl (hunk 1 above)
Comment (iliajie) [reply virtual-server-lib-funcs.pl:12948]: > This could also be a separate PR

Also agreed! Here is a separate PR https://github.com/virtualmin/virtualmin-gpl/pull/904.

File: workbench.cgi (hunk 2 above)
Comment (iliajie) [reply workbench.cgi:10]: Using `$p` makes the code more concise and potentially more flexible for future changes.

File: workbench.cgi (hunk 3 above)
Comment (iliajie) [reply workbench.cgi]: That’s right! There’s no need for this complexity. Fixed [here](https://github.com/virtualmin/virtualmin-gpl/pull/896/commits/750ad2ff7f68044f46921027d6426c805ebfdf89).

File: workbench.cgi (hunk 2 above)
Comment (jcameron) [reply workbench.cgi:10]: No it's just confusing, instead you should just use the string "pro"

File: virtual-server-lib-funcs.pl (hunk 1 above)
Comment (jcameron) [reply virtual-server-lib-funcs.pl]: Thanks!

File: edit_phpmode.cgi (hunk 4 above)
[the remark this answers is not in the document]
Comment (jcameron) [reply edit_phpmode.cgi]: Thanks!

**Comment by shoulders** [comment]
@iliajie This is a great pro feature, does `Virtualmin.com` need a blog to announce features like this? 😄, I only found it by randomly being nosey.

### Server Template Section

A future feature for this, I have not put too much thought into it though, maybe this could be complimented by a server template section so you can setup WordPress automatically and then you could push the fact that Virtualmin can handle preconfigured WebApps rather than just basic websites.

One crazy feature, for the server template, you could have it so the you can configure in the server template to pre-install various plugins such as wordfence, W3C Total cache, maybe this can be done with direct URLs tot he wordpress repo or that a user can upload them to Virtualmin (these would need to be included in the backup).

- The plugins do not need to be configured but just having them pre-installed saves a lot of work
- Comemrcial plugins will not be available via URL

**Comment by iliajie** [comment]
> does `Virtualmin.com` need a blog to announce features like this? 😄, I only found it by randomly being nosey.

Yes, we will make an announcement about it in the forum when it's released.

>  I have not put too much thought into it though, maybe this could be complimented by a server template section

Templates already provide an ability to preinstall web apps with "Default web app installers" page.

File: workbench.cgi (hunk 2 above)
Comment (iliajie) [reply workbench.cgi]: Well, okay! Fixed here https://github.com/virtualmin/virtualmin-gpl/pull/896/commits/75d68bf150b8b3bca4edbcd327a64d039c709e11.

**Comment by shoulders** [comment]
>Templates already provide an ability to preinstall web apps with "Default web app installers" page.

I was meaning to take advantage of some of the advanced features of your `WordPress WorkBench` but all configured via the server template eg: enable fail2ban, enable backups, use this theme etc... so a server template can fully configure WordPress upon install.

**Comment by pixel-paul** [comment]
Being able to install a list of plugins at install would be very useful - kind of what I was eluding to with the mu-plugins.

**Comment by iliajie** [comment]
> @shoulders but all configured via the server template eg: enable fail2ban, enable backups, use this theme etc
>> @pixel-paul Being able to install a list of plugins at install would be very useful 

This sounds like a great improvement for the future; right now, `script_wordpress_kit_apply` doesn’t handle installing WordPress plugins using bundles, but it’s something we could potentially add.

Even though I didn’t want to include plugin management in the WP Workbench UI, I wouldn’t mind adding some new `script-workbench` CLI commands, as it wouldn’t be difficult since `script_wordpress_kit_apply` already handles the heavy lifting.

File: backups-lib.pl [hunk 5]
```diff
@@ -6934,5 +6934,29 @@ sub expand_glob_to_files
...
+# Runs an rsync command to sync source to target with optional extra parameters.
+# If $d is provided, the command is run as the domain owner. The function
+# returns the output of the rsync command, and in list context, it also returns
+# the exit status and the command run.
+sub sync_directories
+{
+my ($source, $target, $rsync_flags, $d) = @_;
+my $rsync = &has_command("rsync");
+my $emsg = &text('restore_ersync', 'rsync');
+return (wantarray ? ($emsg, 1) : 1) if (!$rsync);
```
Comment (jcameron) [inline backups-lib.pl:6947]: You probalby don't need to brackets around `(wantarray ? ($emsg, 1) : 1)`

Comment (iliajie) [reply backups-lib.pl:6947]: Maybe not, but I prefer to keep it because of the following _if_ statement.

**Comment by aryanrajseo** [comment]
i checked all the screenshot. wow and amazing. this is really valuable. thank you for bringing this.

**Comment by Netmees** [comment]
hi, workbench need an improvement asap, when you modify password for a database it does not work anymore so it must be setup a sync action for keep the module working on virtuaserver

**Comment by iliajie** [comment]
> when you modify password for a database

How do you modify password for a database exactly?

**Comment by Netmees** [comment]
excuseme... the user password of the database

**Comment by Netmees** [comment]
from user settings on webmin this its not syncroniced with the script

**Comment by iliaross** [comment]
> from user settings on webmin this its not syncroniced with the script

Please open a new ticket if something still doesn't work for you with the latest version 1.0.8 update.
